Rebase config manager

IRCody opened the pull request:
Fixes #

Summary of changes:
- Adds control.config grpc interface.
- Integrates this interface with rest

Testing done:
- manual
- tests in control/config_proxy_test.go

@intelsdi-x/snap-maintainers

geauxvirtual commented on control/config_proxy.go:
```diff
@@ -0,0 +1,104 @@
+http://www.apache.org/licenses/LICENSE-2.0.txt
+
+
+Copyright 2015 Intel Corporation
```
We should probably be a bit more diligent on setting the year (not like it actually matters from what I have read about having an actual copyright line in your source code) to the year the code landed, so `2016` for anything new.

geauxvirtual commented on mgmt/rest/client/client_func_test.go:
```diff
@@ -100,9 +100,10 @@ func startAPI() (string, error) {
-	r.BindConfigManager(c.Config)
 	r.BindMetricManager(controlClient)
 	r.BindTaskManager(client)
+	configClient, _ := control.NewConfigClient(c.Config.ListenAddr, c.Config.ListenPort)
```
We should catch the error here and kill running the tests if we can't create a client for some reason.

IRCody replied:
Agreed.

geauxvirtual commented on mgmt/rest/rest_func_test.go:
```diff
@@ -472,8 +472,8 @@ func startAPI(cfg *mockConfig) *restAPIInstance {
 	r.BindMetricManager(controlClient)
-
-	r.BindConfigManager(c.Config)
+	configClient, _ := control.NewConfigClient(c.Config.ListenAddr, c.Config.ListenPort)
```
Same here, catch any error and handle it.

geauxvirtual commented on scheduler/workflow.go:
```diff
@@ -315,8 +315,10 @@ func (s *schedulerWorkflow) Start(t *task) {
 
-	if len(errors) != 0 {
-		t.RecordFailure(j.Errors())
+	if len(errors) != 0 && len(j.Errors()) > 0 {
```
Not sure why there is a change in scheduler for this config rpc change. This also seems to be partially reverting a current commit in Master.

IRCody replied:
This line was changed to avoid panic's in the case where len(errors) != 0 but len(j.Errors()) == 0. Overwriting the new changes was not intentional.

geauxvirtual replied:
Can we create an issue on this? It seems with Collect Jobs, we check the return `errors` but then pass `j.Errors()` into `RecordFailures` but for process and publish jobs, we pass `errors` into RecordFailures.

IRCody replied:
I created #858 to bring this up. I'll remove these modifications from this PR.

IRCody commented:
@geauxvirtual, @jcooklin: Updated this a few days ago to address the comments. Would appreciate more feedback on the changes.

geauxvirtual commented on control/config_proxy.go:
```diff
@@ -0,0 +1,104 @@
+	Type := int(r.PluginType)
+	Name := r.Name
+	Version := int(r.Version)
+	result_node := cfg.cfg.GetPluginConfigDataNode(core.PluginType(Type), Name, Version)
```
Maybe from an aesthetics view, change `(cfg *configProxy)` to `(cp *configProxy)` or `(p *configProxy)` in order to not have `cfg.cfg` in the code to make a little clearer.

geauxvirtual commented on control/config_proxy.go:
```diff
@@ -0,0 +1,104 @@
+	Name := r.Name
+	Version := int(r.Version)
+	result_node := cfg.cfg.GetPluginConfigDataNode(core.PluginType(Type), Name, Version)
+	reply.Node, _ = nodeToJSON(&result_node)
```
We should always handle any error returned from JSON Marshalling. It's better to have a few lines of extra code to handle the error versus trying to troubleshoot a corner case of something not working that we expect to work.

IRCody replied:
Thanks for catching that. It is probably better to just not use JSON as an intermediary since it would remove the chance for these errors.

The reason for using JSON was that cdata.ConfigDataNode has a map[string]ConfigValue. ConfigValue is an interface which makes it harder to represent as a protobuf message. I can represent it as a protobuf message instead and provide helper functions to go from node->proto->node to make it simple to use. I think this would end up being better and should probably live in the common proto so it can be re-used.

geauxvirtual replied:
I don't have a problem either way, just as long as we're consistent for everything.

IRCody commented:
@geauxvirtual: Updated to use a new ConfigMap protobuf message defined in internal/common instead passing json as bytes for ConfigDataNode. Added as a separate commit to make it easier to review the changes since it touched many files but plan on rebasing it before merging.

geauxvirtual commented:
LGTM, merging in.
